src/Network/AWS/S3/StreamingUpload.hs

```diff
@@ -85,74 +92,107 @@ See the AWS documentation for more details.
+    startUpload = do
+      multiPartUpload <- lift $ send multiPartUploadDesc
+      when (multiPartUpload ^. cmursResponseStatus /= 200) $
+        fail "Failed to create upload"
+      lift $ logStr "\n**** Created upload\n"
+
+      let Just upId = multiPartUpload ^. cmursUploadId
+          bucket    = multiPartUploadDesc  ^. cmuBucket
+          key       = multiPartUploadDesc  ^. cmuKey
+
+      handleC (cancelMultiUploadConduit bucket key upId) $ 
+        concurrentMapM_ 10 3 (multiUpload bucket key upId) 
+        .| (finishMultiUploadConduit bucket key upId)
+    
+    multiUpload :: (MonadUnliftIO m, MonadAWS m, MonadFail m, MonadResource m) =>
+                   BucketName
+                -> ObjectKey
+                -> Text
+                -> (Int, ByteString)
+                -> m (Maybe CompletedPart)
+    multiUpload bucket key upId (partnum, buffer) = do
+      res <- liftAWS $ send $ uploadPart bucket key partnum upId (toBody $ HashedBytes (hash buffer) buffer)
+      when (res ^. uprsResponseStatus /= 200) $ fail "Failed to upload piece"
+      logStr $ printf "\n**** Uploaded part %d" partnum
+      return $ completedPart partnum <$> (res ^. uprsETag)
+    
+    -- collect all the parts
+    finishMultiUploadConduit :: (MonadUnliftIO m, MonadAWS m, MonadFail m) =>
+                             BucketName
+                             -> ObjectKey
+                             -> Text
+                             -> ConduitT
+                                  (Maybe CompletedPart)
+                                  Void
+                                  m
+                                  (Either (AbortMultipartUploadResponse, SomeException) CompleteMultipartUploadResponse)
+    finishMultiUploadConduit bucket key upId = do
+      parts <- sinkList
+      res <- lift $ send $ completeMultipartUpload bucket key upId & cMultipartUpload ?~ set cmuParts (sequenceA (fromList parts)) completedMultipartUpload
+      return $ Right res
+
+    -- in case of an exception, return Left
+    cancelMultiUploadConduit :: (MonadUnliftIO m, MonadAWS m, MonadFail m) =>
+                                BucketName
+                             -> ObjectKey 
+                             -> Text 
+                             -> SomeException
+                             -> ConduitT
+                                  i
+                                  Void
+                                  m (Either (AbortMultipartUploadResponse, SomeException) CompleteMultipartUploadResponse)
+    cancelMultiUploadConduit bucket key upId exc = do
+      res <- lift $ send $ abortMultipartUpload bucket key upId
+      return $ Left (res, exc)
+
+    -- split stream in chunks of chunkSize
+    chunkConduit :: (MonadUnliftIO m, MonadAWS m, MonadFail m, MonadResource m, PrimMonad m) =>
+                    ConduitT ByteString (VS.Vector Word8) m ()
+    chunkConduit =
+      vectorBuilder (fromIntegral chunkSize) $ \yieldByte ->
```

Comment:
Looking at the source for vectorBuilder, it looks like this won't send anything downstream until all data has been send to this conduit, which would force everything to remain in memory! See the ` Prelude.mapM_ yield vs` in its definition to see where it actually yields the vectors. If I'm right about this then we definitely can't use this as it defeats the purpose of the package.

Reply:
My understanding is that it will force only up to the `chunkSize`, which was the case with the existing code as well. It should be more efficient (less CPU time) building up that `buffer` before it's sent downstream.

Reply:
Take a look at the source for `vectorBuilder`, you'll see that it holds onto all the vectors until no more data is coming from upstream and then yields all the vectors it created, which is definitely unexpected. I made [an issue](https://github.com/snoyberg/mono-traversable/issues/189) on the conduit repo about it, and it looks like I'm right about its behaviour. This is almost certainly where your memory leak is coming from.

The current code spends its time only passing around pointers to existing bytestrings to avoid copying as much as possible. I'm not sure I agree that this could possibly be more efficient than building ups a DList of bytestrings (this is only being done to avoid doing a `reverse` over a list of bytestrings once the chunk size is reached). I was pretty careful to avoid copying bytestring data at all if I could, so that it can be written directly to the HashedBody (see `performUpload`). There's a possibility that this will hold onto more bytestring data than necessary if the byte strings being sent from upstream are slices of much larger bytestrings.

Reply:
Hmm looking at vectorBuilder again, I may be wrong about how it works, it's definitely a confusing piece of code!

Reply:
Yeah vectorBuilder definitely holds onto the entire input its sent before yielding anything.

Reply:
Are you sure? I'm using it for copying 2GB through conduit in another context and memory is constant (up to a few dozen MB).

Reply:
@axman6 I'd appreciate the feedback :)

Reply:
Yeah it looks like I was wrong about the behaviour of vectorBuilder. I'm still not at al convinced by this part of the change though, it pretty much cannot be more efficient than using DList and not copying any data, it turns the algorithm from O(n) where n is the number of ByteStrings sent into O(n) where n is the number of bytes sent, and can only really have benefits if someone is sending lots of bytestrings which are small slices of much larger ones.

I'll have to have a look this weekend, there's a lot I like about how this has been rewritten, and the concurrency changes are quite nice in particular. I've never been happy with the spaghetti structure of the code and this at least gives me a good way to clean it up.

Reply:
It was inspired by https://github.com/blitzcode/conduit-chunked, But it's currently slower as there is a leak somewhere.